src/nekrs_driver.cpp

```diff
@@ -0,0 +1,229 @@
+#include "enrico/nekrs_driver.h"
+#include "enrico/error.h"
+#include "gsl.hpp"
+#include "iapws/iapws.h"
+#include "libP/include/mesh3D.h"
+#include "nekrs.hpp"
+
+#include <algorithm>
+#include <dlfcn.h>
+
+namespace enrico {
+NekRSDriver::NekRSDriver(MPI_Comm comm, pugi::xml_node node)
+  : HeatFluidsDriver(comm, node)
+{
+
+  if (active()) {
+    // See vendor/nekRS/src/core/occaDeviceConfig.cpp for valid keys
+    setup_file_ = node.child_value("casename");
+    nekrs::setup(comm, 0, 0, 0, "", setup_file_, "", "");
+
+    open_lib_udf();
+
+    // Check that we're running a CHT simulation.
+    err_chk(nekrs::cht(),
+            "NekRS simulation is not setup for conjugate-heat transfer (CHT).  "
+            "ENRICO must be run with a CHT simulation.");
+
+    // Local and global element counts
+    n_local_elem_ = nekrs::mesh::nElements();
+    std::size_t n = n_local_elem_;
+    MPI_Allreduce(
+      &n, &n_global_elem_, 1, get_mpi_type<std::size_t>(), MPI_SUM, comm_.comm);
+
+    poly_deg_ = nekrs::mesh::polyDeg();
+    n_gll_ = (poly_deg_ + 1) * (poly_deg_ + 1) * (poly_deg_ + 1);
+
+    x_ = nekrs::mesh::x();
+    y_ = nekrs::mesh::y();
+    z_ = nekrs::mesh::z();
+    element_info_ = nekrs::mesh::elementInfo();
+
+    // rho energy is field 1 (0-based) of rho
+    rho_cp_ = nekrs::rhoCp();
+
+    temperature_ = nekrs::scalarFields();
+
+    // Construct lumped mass matrix from vgeo
+    // See cdsSetup in vendor/nekRS/src/core/insSetup.cpp
+    mass_matrix_.resize(n_local_elem_ * n_gll_);
+    for (gsl::index e = 0; e < n_local_elem_; ++e) {
+      for (gsl::index n = 0; n < n_gll_; ++n) {
+        mass_matrix_[e * n_gll_ + n] =
+          nekrs::mesh::vgeo()[e * n_gll_ * nekrs::mesh::nVgeo() + JWID * n_gll_ + n];
+      }
+    }
+
+    init_displs();
+  }
+  comm_.Barrier();
+}
+
+void NekRSDriver::init_step()
+{
+  auto min = std::min_element(localq_->cbegin(), localq_->cend());
+  auto max = std::max_element(localq_->cbegin(), localq_->cend());
+}
+
+void NekRSDriver::solve_step()
+{
+  const auto start_time = nekrs::startTime();
+  const auto final_time = nekrs::finalTime();
+  const auto dt = nekrs::dt();
+
+  time_ = start_time;
+  tstep_ = 1;
+
+  while ((final_time - time_) / (final_time * dt) > 1e-6) {
+    nekrs::runStep(time_, dt, tstep_);
+    time_ += dt;
+    nekrs::udfExecuteStep(time_, tstep_, 0);
+    ++tstep_;
+  }
+  // LOOK HERE!!!
```

Comment:
???

Reply:
I'll get rid of that.  It was re: the call `copyToNek`
```suggestion
```